### ISSUE TITLE
Attaching/Detaching One/Many Objects

### ISSUE THREAD

**Issue by DellanX** [dipscope/JsonApiEntityProvider.TS#7]
**Is your feature request related to a problem? Please describe.**
I'd like to be able to quickly Attach/Detach related objects.
This might be something that already exists; however, I don't see how to execute this functionality.

**Describe the solution you'd like**
Ideally, I could just call on an entitySet an attach method such as:
```ts
entitySet.attach(parent, child, relationship)
```
Or better yet, calling `entitySet.save` would be smart enough to attach the related objects
My concern here is amending/replacing on attachMany. Since we might not have included the full list of related objects.

**Describe alternatives you've considered**
Aside from Manually communicating with the server to create these objects, I could create the pivot model via the entity provider; however, both of these options are sloppy.

**Additional context**
[JSON API updating related objects](https://jsonapi.org/format/#crud-updating-relationships)

**Comment by dpimonov** [comment]
Hi DellanX,

Thanks for your issue! 

Unfortunately there is no any change tracking in this library. You can simply add your relations directly to the model properties and they will be serialized during the request. Something like:

```
parent.children.push(child);
```

The library simply just creates a graph from JSON. You can change this graph as you want and then call save method which serializes it back based on your configuration.

**Comment by DellanX** [comment]
Ohhh, that makes sense. Does this serialization execute the appropriate Jsonapi calls? Or is it just a single update call? My main concern again is that the M-M relationships are established correctly

**Comment by dpimonov** [comment]
Library uses only main routes on requests. There are basically 2 options:

1) Declare M-M relationships explicitly. For example each `User` might have application options and each `Option` belongs to many users. There should be `UserOption` entity with its own endpoint. Then you can use save operations the same way as usual without relationship routes.

2) Use entity properties directly to update relationships in case when you have to update multiple relationships. For this you have to set this option to true when setup entity provider. It is false by default. 

![image](https://github.com/dipscope/JsonApiEntityProvider.TS/assets/15626139/ca188bfb-d7fd-41e5-b4b4-92cfa20d79a4)

Then you can follow [this part of spec](https://jsonapi.org/format/#crud-updating-resource-relationships) to update your relationships. Basically when option is enabled - serializer will also include all entity relationships on save requests.

**Comment by DellanX** [comment]
Okay, I understand pretty well now (Unfortunately, where I am using Pagination, so I'll need to be able to use a POST on the nested endpoint)

On a further note, is there any plans to add support for the relationship endpoints, as was mentioned in #2?
I can help if desired, as this is the only feature I need to be added to switch to this library, and this library is far superior to my current setup.

**Comment by dpimonov** [comment]
Can you describe your use case in details please? The library provides pagination, filtering, inclusion and sorting api to get the data in many ways. It provides options to save relationships together with entity. So far I do not clearly understand why nested endpoints are required.

```
On a further note, is there any plans to add support for the relationship endpoints.
```

There might be depending from a use case. If there is really something missing - help is appreciated. :)

**Comment by DellanX** [comment]
There are many parts to this.
In my code, let's say that there are chats, and in that chats there are messages.
For the sake of this issue, these are attached M-M. So we have the following endpoints:
```txt
/chats
/chats/:id
/messages
/chats/:id/messages
```

I'd need to be able to do:
1. Query paginated messages from the chat.
2. Add a message to the chat.

Now, here is where it gets tricky, and why I want the nested endpoints.
On the server, the authentication policy for `/chats` is different from the policy for `/chats/:id/messages`

I don't NEED the nested endpoints but do want them.
I can achieve this via inclusion and filtering, the main headaches I am running into is:
1. Authorization Policy issues (Server-side problem, which I can work at)
2. Adding a single Message to the chat.

**Comment by dpimonov** [comment]
Basically for many-to-many relationships it is always better to have a pivot entity like `ChatMessage`. It is more future proof. Later some additional properties related to the link it self might be added. But I've got a point and use case. 

I'm going to think how it can be added. First of all this library is a plugin for `EntityStore` which is basically `ORM`-like abstraction and from api point of view I want to keep it as simple as possible and don't want to extend its api with library specific functions. However is it by design provides extension points for different parts which we can use.

So far to get use of relationships especially many-to-many the first idea is to use library related entity collections like it is done during pagination. 

```
RelationshipEntityCollection<TParentEntity, TEntity> : EntityCollection<TEntity>
```

As there are many ways to work with relationships in spec and I want to keep this behaviour configurable. Such collections might be created by default during serialization/deserialization or be used during entity definitions. When one calls save methods - such collection might be treated differently and we can get use of relationship links through the parent entity. 

So if library sees that save method is called on relationship entity collection then a request to `parent/:id/relationsip/entity` is sent using relationship links as describe in spec. However how to treat one-to-many relationships is still an open question. But... for your concrete case the code might look like this:

```
const messageEntitySet = entityStore.messages;
const chatMessages = new JsonApiRelationshipEntityCollection(chat, messages);

messageEntitySet.add(chatMessages);
messageEntitySet.update(chatMessages);
messageEntitySet.remove(chatMessages);
```

Also this might solve an issue with pagination as if one defines an entity like:

```
export class Chat
{
    @Property(JsonApiRelationshipEntityCollection, [Chat, Message]) 
    public messages: JsonApiRelationshipEntityCollection<Chat, Message>;
}
```

The the whole serialization process might be applied and he can get pagination and saving stuff out of the box with minimum efforts. 

This is just a raw idea right now. If this will be sufficient - we can think about implementation details as in this case I would like to cover the whole relationship api and not only many-to-many use case. Tell me what you think.

**Comment by DellanX** [comment]
Actually, I may be able to circumvent a bunch of the issues by doing things by the Pivot Model as said.
In that case, I have to worry about handing two 1-M requests.

Let's make the example a bit more complicated. Say you have 3 models, and 1 pivot model:
* Chat
* Message
* User
* ReadReceipt

Chats has Messsages as a 1-M relationship
Messages have Users as a M-M using ReadReceipts as the Pivot Model

We would then have the following routes:
```text
/chats/:id
/chats/:id/relationships/messages
/messages/:id/relationships/users
/messages/:id/relationships/read-receipts
/users/:id/relationships/read-receipts
```

While yes, I could have some `/read-receipts` routes, I would have to modify my server to handle permission management based upon the parameters provided for both `userId` and `messageId`. An inconvenience, but definitely possible.

The best option, in my mind, would be along what you're stating.
I didn't quite understand the entitySet logic, but figure you're probably more onto the right idea.
For now, I just used `RelationshipApi` to represent the container object in its place.

```ts
// Adding Messages
const chatMessageSet = new RelationshipApi<Message>(chat)

chatMessageSet.add(message)
chatMessageSet.update(messages)
chatMessageSet.remove(message)

// Adding Read Receipts
// Read Receipt would have userId populated before sent
const messageReadReceiptSet = new RelationshipApi<ReadReceipt>(message)

messageReadReceiptSet.add(readReceipt)
messageReadReceiptSet.update(readReceipts)
messageReadReceiptSet.remove(readReceipt)

// Adding Users via Read Receipts
// The EntityStore would need to create the ReadReceipt manually somehow...
const messageUserSet = new RelationshipApi<Message, ReadReceipt>(message)

messageUserSet.add(user)
messageUserSet.update(users)
messageUserSet.remove(user)

```

**Comment by dpimonov** [comment]
Thanks for your comment! I think about possible ways to solve an issue. I have some free time only on the weekend. If I find something which fits - I will return to you with concrete proposal.

**Comment by dpimonov** [comment]
So, I've implemented all relationship api according to the spec and committed to the main branch. Here is how to use it:

```
const chatSet = entityStore.chatSet;
const chat = chat;
const chatMessagesRelationship = entityProvider.createJsonApiToManyRelationship(chatSet, chat, c => c.messages);

chatMessagesRelationship.add(message);
chatMessagesRelationship.bulkAdd(messages);
chatMessagesRelationship.update(message);
chatMessagesRelationship.bulkUpdate(messages);
chatMessagesRelationship.remove(message);
chatMessagesRelationship.bulkRemove(messages);
```

Working with to-one relations is more or less similar. I will need your help with implementing tests for both to-many and to-one relationship classes and new section in the docs. I've tested only add function and created further test scenarios. All other parts are so far untested. You can create a test server by simply runing backend project and tests by following already implemented test scenarios. After everything is done and documented - we can make a new release. I am available for any questions.

**Comment by DellanX** [comment]
Neet! I like the top-level look at the functionality.

It may be a few days before I am able to setup the experiment and create the tests but will do so when I get a chance.

**Comment by dpimonov** [comment]
Sure! Take your time.

**Comment by DellanX** [comment]
How difficult would it be to add support for querying the relationships as well?

Was designing an experiment, and realized that if we have an object representing the relationship already, it'd be nice to just issue a `get` or `find` command. My current codebase does this by `include` blocks; however, when it comes to a chat, it'd be nice to apply pagination to messages.

(My actual use case requires me to be able to apply filters on the relationships, as well as pagination)

**Comment by DellanX** [comment]
^Disregard, it looks to me like you've already implemented `findAll`

I'll come up with some tests, add them to your spec files, and submit a PR.
Plus, I'll ping back here if there are any other ideas I come up with.

Thanks for the work on this, it looks amazing! 😄

**Comment by dpimonov** [comment]
You are welcome! 😊

Basically with relationship api I included everything which can be supported easily. This covers `findAll` method. With a `find` method it is a little bit more complicated when it comes to relationships as filtering api should be used internally. So for now I simple ignore it but we can implement it also if really required. Anyway one can filter relationships with `filter` + `findOne` method and get the same result.

I am looking forward for your PR. If you will need any help with running C# backend - tell me. Thanks for the help! 😉

**Comment by DellanX** [comment]
Hi, I may have something worth mentioning, which you may already be aware of.

There are a few ways to query the relationships via JsonApi in the spec.
Usually there is a self-link, and a related link.
* Self Link `parent/relationships/child` returns the types and ids of the relationships. Not the attributes.
(Resource Identifier Objects)
* Related Link `parent/child` returns the types, ids and attributes of the relationships.
(The objects schema including Fields)

For updating the relationships, the self-link shines, because you can modify the relationships with only the type and id information.

But for fetching the information, it'll need to use the related link.
(Otherwise, two queries will be needed, one to get the IDs, another to get the data)

I tried to find the actual clauses in the spec; however, link objects are a bit difficult to navigate. Here is a good [stack overflow](https://stackoverflow.com/questions/32522212/jsonapi-difference-between-self-and-related-in-a-links-resource) page on the matter

**Comment by DellanX** [comment]
Submitted a PR for the tests.
The feature may still need some work, as two of the tests are failing.
* FindAll not including Fields
* Include not attaching related object

Otherwise, it's working pretty well!
(If I end up with time to work on fixing the two issues above, I'll post here)

Now that I am able to run your development setup, I may also add a few improvements. But will create a draft PR for those separately from the test PR.

**Comment by dpimonov** [comment]
Included in v2.1.0.
